test/instrumentation/modules/aws-sdk/sqs.js

```diff
@@ -0,0 +1,409 @@
+
+        const spanHttp = data.spans[1]
+        t.equals(spanHttp.type, 'external', 'second span is for HTTP request')
+        t.end()
+      })
+      agent.startTransaction('myTransaction')
+      const sqs = new AWS.SQS({ apiVersion: '2012-11-05' })
+      const params = getParams('deleteMessageBatch', listener.address().port)
+      sqs.deleteMessageBatch(params, function (err, data) {
+        t.error(err)
+        agent.endTransaction()
+        listener.close()
+      })
+    })
+  })
+
+  test.test('API: receiveMessage', function (t) {
+    const app = createMockServer(
+      getXmlResponse('receiveMessage')
+    )
+    const listener = app.listen(0, function () {
+      resetAgent(function (data) {
+        t.equals(data.spans.length, 2, 'generated two spans')
+
+        const spanSqs = data.spans[0]
+        t.equals(spanSqs.name, 'SQS POLL from our-queue', 'SQS span named correctly')
+        t.equals(spanSqs.type, 'messaging', 'span type set to messaging')
+        t.equals(spanSqs.subtype, 'sqs', 'span subtype set to sqs')
+        t.equals(spanSqs.action, 'poll', 'span action matches API method called')
+        t.equals(spanSqs.context.destination.service.type, 'messaging', 'messaging context set')
+
+        const spanHttp = data.spans[1]
+        t.equals(spanHttp.type, 'external', 'second span is for HTTP request')
+        t.end()
+      })
+      agent.startTransaction('myTransaction')
+      const sqs = new AWS.SQS({ apiVersion: '2012-11-05' })
+      const params = getParams('receiveMessage', listener.address().port)
+      sqs.receiveMessage(params, function (err, data) {
+        t.error(err)
+        agent.endTransaction()
+        listener.close()
+      })
+    })
+  })
+
+  test.test('API: receiveMessage no transaction', function (t) {
+    const app = createMockServer(
+      getXmlResponse('receiveMessage')
+    )
+    const listener = app.listen(0, function () {
+      resetAgent(function (data) {
+        t.equals(data.spans.length, 0, 'no spans without a transaction')
+        t.end()
+      })
+
+      const sqs = new AWS.SQS({ apiVersion: '2012-11-05' })
+      const params = getParams('receiveMessage', listener.address().port)
+      sqs.receiveMessage(params, function (err, data) {
+        t.error(err)
+        listener.close()
+      })
+    })
+  })
+
+  test.test('API: sendMessage without a transaction', function (t) {
+    const app = createMockServer(
+      getXmlResponse('sendMessage')
+    )
+    const listener = app.listen(0, function () {
+      resetAgent(function (data) {
+        t.equals(data.spans.length, 0, 'no spans without a transaction')
+        t.end()
+      })
+      const sqs = new AWS.SQS({ apiVersion: '2012-11-05' })
+      const params = getParams('sendMessage', listener.address().port)
+      sqs.sendMessage(params, function (err, data) {
+        t.error(err)
+        listener.close()
+      })
+    })
+  })
+
```

Comment:
Would it be possible (I'm not sure with the mocking here) to test the promise-y AWS SDK usage (https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/using-promises.html)?

Reply:
Good call -- using the promise API was generating HTTP spans, but not SQS spans.  The promise API bypasses the `send` method and instead [adds directly to that state-machine-like-thing](https://github.com/aws/aws-sdk-js/blob/3691e826f4e785d7ead9763f10d58fa098a4078d/lib/request.js#L797) that's under the hood.  I've added a wrapper for `promise` as well to catch this case.